components/logins/src/db.rs

```diff
@@ -1235,4 +1268,108 @@ mod tests {
         db.delete(login.guid_str()).unwrap();
         assert!(&unique_login_check.is_ok())
     }
+
+    fn check_matches(db: &LoginDb, query: &str, expected: &Vec<&str>) {
+        let mut results = db
+            .get_by_base_domain(query)
+            .unwrap()
+            .into_iter()
+            .map(|l| l.hostname)
+            .collect::<Vec<String>>();
+        results.sort();
+        let mut sorted = expected.clone();
+        sorted.sort();
+        assert_eq!(sorted, results);
+    }
+
+    fn check_good_bad(
+        good: Vec<&str>,
+        bad: Vec<&str>,
+        good_queries: Vec<&str>,
+        zero_queries: Vec<&str>,
+    ) {
+        let db = LoginDb::open_in_memory(Some("testing")).unwrap();
+        for h in good.iter().chain(bad.iter()) {
+            db.add(Login {
+                hostname: (*h).into(),
+                http_realm: Some((*h).into()),
+                password: "test".into(),
+                ..Login::default()
+            })
+            .unwrap();
+        }
+        for query in good_queries {
+            check_matches(&db, query, &good);
+        }
+        for query in zero_queries {
+            check_matches(&db, query, &vec![]);
+        }
+    }
+
+    #[test]
+    fn test_get_by_base_domain_invalid() {
+        check_good_bad(
+            vec!["https://example.com"],
+            vec![],
+            vec![],
+            vec!["invalid query"],
+        );
+    }
+
+    #[test]
+    fn test_get_by_base_domain() {
+        check_good_bad(
+            vec![
+                "https://example.com",
+                "https://www.example.com",
+                "http://www.example.com",
+                "http://www.example.com:8080",
+                "http://sub.example.com:8080",
+                "https://sub.example.com:8080",
+                "https://sub.sub.example.com",
+            ],
+            vec![
+                "https://badexample.com",
+                "https://example.co",
+                "https://example.com.au",
+            ],
+            vec!["example.com"],
+            vec!["foo.com"],
+        );
+        // punycode! This is likely to need adjusting once we normalize
+        // on insert.
+        check_good_bad(
+            vec![
+                "http://😍.com",
+                "http://xn--r28h.com", // punycoded version of the above.
+            ],
+            vec!["http://💖.com"],
+            vec!["😍.com", "xn--r28h.com"],
```

Comment:
nit: this test might start failing when we get around to de-duping on punycode hostnames, I wonder if it would be more robust to have one of these be a subdomain of the other.

Reply:
It might, but maybe not. The point here is to test that the *search string* is also normalized, and if that keep happening then the test should not fail. If we decide to *not* do that later, it's good to have that explicitly called out.

FWIW, the exact same issue exists with the ipv6 test, where the same address in 2 different formats is used.

So if it's ok with you, I think it's best left alone.

Reply:
totally fine by me 👍